contracts/MultiRewards.sol

```diff
@@ -535,28 +622,85 @@ contract MultiRewards is ReentrancyGuard, Pausable {
         rewardData[_rewardsToken].rewardsDuration = _rewardsDuration;
         emit RewardsDurationUpdated(_rewardsToken, rewardData[_rewardsToken].rewardsDuration);
     }
+    */
 
     /* ========== MODIFIERS ========== */
 
-    modifier updateReward(address account) {
-        for (uint i; i < rewardTokens.length; i++) {
-            address token = rewardTokens[i];
-            rewardData[token].rewardPerTokenStored = rewardPerToken(token);
-            rewardData[token].lastUpdateTime = lastTimeRewardApplicable(token);
-            if (account != address(0)) {
-                rewards[account][token] = earned(account, token);
-                userRewardPerTokenPaid[account][token] = rewardData[token].rewardPerTokenStored;
+    function _updateReward() internal {
+        IHypervisor(stakingToken).getReward();
+        for (uint i; i < rewardTokens.length; i ++) {
+            address rewardToken = rewardTokens[i];
+            if (totalStakes > 0) {
+                RewardData storage r = rewardData[rewardToken];
+                uint256 currentBalance = IERC20(rewardToken).balanceOf(address(this));
+                uint256 diff =  currentBalance - r.amount;
+                r.lastTimeUpdated = block.timestamp;
+                r.rewardPerToken += diff * 1e50 / totalStakes;
+                r.amount = currentBalance;
             }
         }
-        _;
     }
 
     /* ========== EVENTS ========== */
 
-    event RewardAdded(uint256 reward);
+
     event Staked(address indexed user, uint256 amount);
-    event Withdrawn(address indexed user, uint256 amount);
+    event Unstake(address indexed user, uint256 amount);
     event RewardPaid(address indexed user, address indexed rewardsToken, uint256 reward);
-    event RewardsDurationUpdated(address token, uint256 newDuration);
+
     event Recovered(address token, uint256 amount);
+
+    /********************** Errors ***********************/
+    error AddressZero();
+    error InvalidBurn();
+    error InsufficientPermission();
+    error ActiveReward();
+    error InvalidAmount();
+
+    // Extra storage
+
+    /// @notice Addresses approved to call mint
+    mapping(address => bool) public managers;
```

Comment:
Not too sure what the `managers` are for, from what I can tell the only use for this is to control who can add a `rewardToken` via `addReward`

Reply:
I think so too. We can decide not to have them. Or keep them - either way is fine

Reply:
I think we should keep them - less differences from Gamma's contract and we may use them later